[mlir][tensor][linalg] Move Pack/Unpack Ops to Linalg

banach-space opened the pull request:
This is merely moving code around, no new functionality is added.

Given the size of this change, I've split it into 5 self-contained commits:
* [PATCH 1](https://github.com/llvm/llvm-project/pull/123902/commits/e852e68235adfcfc0f0d8e3ccf054442c06886a4): **Duplicate the definitions** of `tensro.pack`+`tensor.unpack` as `linalg.pack`+`linalg.unpack` (TableGen + CMake + verifiers + builders).
  * This patch only adds code.

* [PATCH 2](https://github.com/llvm/llvm-project/pull/123902/commits/b9d0e09e635bcf0ea617b9bdda71a5fd14f22f5f): **Port tests from “ops.mlir” and "invalid.mlir"**
  * This patch only moves the most basic tests.

* [PATCH 3](https://github.com/llvm/llvm-project/pull/123902/commits/71da9f983219e4cae91d9ce491eafd87e5b1d745): **Port all the outstanding tests**.
  * Big change, updates the corresponding testing logic and moves all the relevant patterns + helpers.

* [PATCH 4](https://github.com/llvm/llvm-project/pull/123902/commits/947558d3a9a4777dfc2f3d69b7d63d653596b9b3): **Delete what’s left.**
  * This patch only deletes code.
  
CONTEXT:
This change was discussed in the following RFC:
* https://discourse.llvm.org/t/rfc-move-tensor-pack-and-tensor-unpack-into-linalg

--------
**Suggestion for reviewers**

I think that the best way to review this is to look at stats, e.g.
* 4515 lines inserted, 4302 lines deleted (sounds about right).

and to make sure that things are indeed just moved around, e.g. [PATCH 2](https://github.com/llvm/llvm-project/pull/123902/commits/ee159c2f725345cffb8bc2ea08f9d040ff0fe4a8) makes this very clear. Sadly, it's a bit tricky to split the largest patch, [PATCH 3](https://github.com/llvm/llvm-project/pull/123902/commits/60d1e17f9de0a690da309e481e8280afa62cdf52).

banach-space committed:
[mlir][tensor][linalg] Move Pack/Unpack Ops to Linalg (3/4)

This is merely moving code around, no new functionality is added.

PATCH 3: Update/move/replace all tests for `tensor.{pack|unpack}` with
identical tests for `linalg.{pack|unpack}`. Updates the testing
infrastructure accordingly and copy all the required transformations.

To help reviewing, below is an overview of non-obvious code moves:

1. Tests from:
  * "mlir/test/Dialect/Tensor/tiling.mlir"
are moved to to:
  * "mlir/test/Dialect/Linalg/transform-op-tile-pack-unpack.mlir"

2. Tests from:
  * "mlir/test/Dialect/Tensor/fold-empty-op.mlir"
are moved to:
  * "mlir/test/Dialect/Linalg/fold-empty-op.mlir"

CONTEXT:
This change was discussed in the following RFC:
* https://discourse.llvm.org/t/rfc-move-tensor-pack-and-tensor-unpack-into-linalg

banach-space committed:
[mlir][tensor][linalg] Move Pack/Unpack Ops to Linalg (1/4)

This is merely moving code around, no new functionality is added.

PATCH 1: Copies `tensor.pack` and `tensor.unpack` as `linalg.pack` and
`linalg.unpack`, respectively. New Ops are defined in
LinalgRelayoutOps.td.

Note, `tensor.pack` and `tensor.unpack` are still present at this point.

CONTEXT:
This change was discussed in the following RFC:
* https://discourse.llvm.org/t/rfc-move-tensor-pack-and-tensor-unpack-into-linalg

banach-space committed:
[mlir][tensor][linalg] Move Pack/Unpack Ops to Linalg (4/4)

This is merely moving code around, no new functionality is added.

PATCH 4: Remove `tensor.{pack|unpack}` and all the associated code (e.g.
transfromations, verifiers, etc).

CONTEXT:
This change was discussed in the following RFC:
* https://discourse.llvm.org/t/rfc-move-tensor-pack-and-tensor-unpack-into-linalg

banach-space committed:
[mlir][tensor][linalg] Move Pack/Unpack Ops to Linalg (2/4)

This is merely moving code around, no new functionality is added.

PATCH 2: To verify the newly added Ops (and to make the subsequent
change smaller), this PR:
1. Moves tests from:
  * "mlir/test/Dialect/Tensor/ops.mlir"
to:
  * "mlir/test/Dialect/Linalg/named-ops.mlir"

2. Moves tests from:
  * "mlir/test/Dialect/Tensor/invalid.mlir"
to:
  * "mlir/test/Dialect/Linalg/invalid.mlir:

In addition, I grouped "invalid" tests for `linalg.pack` and
`linalg.unpack` into two seperate sets (as opposed to mixing them
together).

CONTEXT:
This change was discussed in the following RFC:
* https://discourse.llvm.org/t/rfc-move-tensor-pack-and-tensor-unpack-into-linalg

rengolin commented:
All 5 commits in your proposal are in this PR. It seems this is not just "patch 1". Is the intention that we review all of them here, now?

MaheshRavishankar commented:
I don't know if this is worth doing now. This is just churn and disruptive. We need to have a good story to split tensor dialect, and I don't see a value in the churn of just these two ops right now. I'll leave other stake holders to decide, but I would rather not do this

Cc @qedawkins @Groverkss @stellaraccident

stellaraccident commented:
It sounds like this is going to sit for a few weeks anyway. Agreed that a bit of a more holistic plan would be better, but I do expect that a first step would be these. 

I'll chat with you offline. Maybe we can get that more holistic plan drafted in the meantime. There's never going to be a really good time to churn on this...

MaheshRavishankar commented:
I missed the discourse post on this. Thanks @banach-space for doing this and the due diligence

banach-space commented:
> All 5 commits in your proposal are in this PR. It seems this is not just "patch 1". 

Sorry, I was meant to update this summary. That's now been fixed.

> Is the intention that we review all of them here, now?

What works for you best?

From my experience, some reviewers prefer **big changes**, while others favor many **smaller patches**. Personally, I find multiple smaller patches more manageable. However, in this case, I was concerned about creating a state where we have:

* `tensor.pack` + `tensor.unpack`, and
* `linalg.pack` + `linalg.unpack`

Also, given our current struggles with reviews, I thought that perhaps "less would be more". I'm happy either way - whatever makes things easier for reviewers. 

WDYT?

rengolin commented:
> I'll chat with you offline. Maybe we can get that more holistic plan drafted in the meantime. There's never going to be a really good time to churn on this...

We have discussed this plan throughout the past year, including @MaheshRavishankar and @banach-space, and the original RFC was a response to that discussion.

There was strong consensus on the RFC and a lot of good discussions on the rationale, the holistic plan, and all short of a roadmap. But we all agreed this would be a reasonable first step.

I'm surprised we're still going in circles.

rengolin commented:
> From my experience, some reviewers prefer **big changes**, while others favor many **smaller patches**. Personally, I find multiple smaller patches more manageable. 

I _"prefer"_ less churn, and sometimes that's lots of small patches, sometimes that one giant patch.

> However, in this case, I was concerned about creating a state where we have:
> 
> * `tensor.pack` + `tensor.unpack`, and
> * `linalg.pack` + `linalg.unpack`

Yes, I'm worried about that, too. Do that in between releases and you have a broken compiler. In any case, perhaps we should wait for the new release to branch before merging?

> Also, given our current struggles with reviews, I thought that perhaps "less would be more". I'm happy either way - whatever makes things easier for reviewers.

So, I looked at the code and it looks very much like my work-in-progress branch doing the same thing. I don't think there's a better way of doing these moves, other than creating new operations entirely, which leads to duplication.

The main problem here is that pack/unpack has received a lot of attention in the past few months (thanks, btw!), and the infra around it got quite big.

For us, a single commit would be easier to rebase on top of. I imagine this is true for others, too. So to me, this would be less churn. But we can accommodate multiple patches if others prefer it that way.

qedawkins commented:
> There was strong consensus on the RFC and a lot of good discussions on the rationale, the holistic plan, and all short of a roadmap. But we all agreed this would be a reasonable first step.
>
> I'm surprised we're still going in circles.

Agree, this seems well in line with the RFC, thanks @banach-space!

I think the key consideration here is giving downstream users a chance to prepare for integration before landing. Maybe we give this a couple weeks for visibility after review?

> For us, a single commit would be easier to rebase on top of. I imagine this is true for others, too.

+1 single commit sounds easier for IREE, especially because this is primarily a rename (or at least making sure all patches here land as a group). I appreciate the PR being broken up into multiple patches for review though.

Groverkss commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td:
```diff
@@ -178,6 +178,14 @@ def LinalgConvolutionOpInterface : OpInterface<"ConvolutionOpInterface"> {
   ];
 }
 
+// TODO:
+def LinalgRelayoutOpInterface : OpInterface<"RelayoutOpInterface"> {
+  let description = [{
+    TODO
+  }];
+  let cppNamespace = "::mlir::linalg";
+}
+
```
nit: remove TODO

Groverkss commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -0,0 +1,332 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This is the operation definition file for structured operations on buffers
+// that correspond to underlying library calls (e.g. BLAS).
```
nit: This needs to be updated, looks like it was copied?

banach-space replied:
Good catch! Updated in the latest revision.

Groverkss commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -0,0 +1,332 @@
+// RelayoutOp
+//===----------------------------------------------------------------------===//
+
+class Linalg_RelayoutOp<string mnemonic, list<Trait> traits = []> :
```
Interesting, it looks like this class is designed with only tensor types in mind (based on the traits NoMemoryEffect and the methods). Eventually we should design it to work on buffers also.

Not for this patch though, just a TODO would be good.

rengolin replied:
A side effect of being created inside the tensor dialect.

banach-space replied:
Added a TODO at the top of the file.

Groverkss commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -126,6 +126,28 @@ def ApplyPadVectorizationPatternsOp : Op<Transform_Dialect,
+def ApplyFoldPackUnpackIntoEmptyPatternsOp : Op<Transform_Dialect,
+    "apply_patterns.linalg.fold_pack_unpack_into_empty",
+    [DeclareOpInterfaceMethods<PatternDescriptorOpInterface>]> {
+  let description = [{
+    // TODO:
+  }];
+
```
nit: Remove TODO

banach-space replied:
Fixed in https://github.com/llvm/llvm-project/pull/125823/

Groverkss commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -0,0 +1,332 @@
+      Op<Linalg_Dialect, mnemonic, !listconcat(traits, [
+        DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+        DestinationStyleOpInterface, LinalgRelayoutOpInterface,
+        ConditionallySpeculatable, NoMemoryEffect,
```
I think DPS, tensor only ops should always be Pure. Is there a reason we have them as ConditionallySpeculatable?  Just trying to understand if there is some possibility of undefined behavior i'm missing.

Groverkss replied:
Sure, as long as there is a TODO / issue tracking it, that is enough for me

Groverkss commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -126,6 +126,28 @@ def ApplyPadVectorizationPatternsOp : Op<Transform_Dialect,
+    [DeclareOpInterfaceMethods<PatternDescriptorOpInterface>]> {
+  let description = [{
+    Indicates that operations like tensor.pad and tensor.extract_slice should
+    be folded into tensor.pack and tensor.unpack operations, respectively.
```
nit: tensor.pack/tensor.unpack -> linalg.pack/linalg.unpack

banach-space commented:
_(NOTE: rebased + updated links)_

Thanks for all the comments so far. I will be OOO next week, so here's just a quick summary + next steps.

--- 
**There's preference to land this as a single commit**
* The split into multiple commits is only to facilitate reviewing. All commits will be squashed before merging into `main`.
---

**Step 1:**  I will extract [PATCH 5](https://github.com/llvm/llvm-project/pull/123902/commits/d70f9d5216f16f92812c9d18b8dcb53b4c038169) into a separate PR - thanks for all the comments so far @Groverkss !

**Step 2:** Once this PR is accepted, I will squash all commits and ping IREE (@qedawkins ?) and MLIR-TPP (@rengolin ?) to try cherry-picking to test downstream. 
* Let's allocate approximately two weeks for collecting and addressing feedback, with flexibility to extend if needed.

**Step 3:** If all is well, I will proceed to land the changes 🥳

rengolin commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -483,3 +483,13 @@ PackingMetadata mlir::computePackingMetadata(int64_t packedRank,
   return res;
 }
+
+OpFoldResult mlir::reshapeConstantSource(DenseElementsAttr source,
```
Is the comment not relevant any more?

banach-space replied:
Note that I create a declaration in this header file:
* mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h

That's where the comment got relocated :)

rengolin reviewed (commented):
Review of patch 1, looks good, minor comments.

rengolin reviewed (commented):
patch 2 is trivial. patch 3 is still copying things from tensor to linalg. haven't checked if all that got added on patches 1 & 3 got deleted on patch 4. :)

But mostly mechanical changes anyway, so both look good, too.

rengolin commented:
Patch 4 looks ok, too. Patch 5 looks like WIP? If we move that to followup PR, this PR LGTM (with comments addressed). Thanks!

But, as @qedawkins said, we need to test it locally first before merge, so I won't press the green button right now.

banach-space commented:
**Update (5/2/24)**

> Patch 5 looks like WIP? If we move that to followup PR, this PR LGTM (with comments addressed).

In fact, [PATCH 5](https://github.com/llvm/llvm-project/pull/123902/commits/f350dd269a195ab791c00d7aa7db147bc7b9ce78) is a pre-requisite of this. I moved it into a dedicated PR (that was my original intention):
* https://github.com/llvm/llvm-project/pull/125823

That PR contains the full rationale for `TensorRelayoutOpInterface`/`LinalgRelayoutOpInterface`. Once folks agree that #125823 is the right pre-requisite step, I will push a re-based version of this PR (again, trying to reduce GitHub noise).

Thank you for all the feedback so far!

Groverkss reviewed (commented):
LGTM, Thanks!. I'm going to pass on the review to reviewers testing this patch downstream.

banach-space commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3616,6 +3604,78 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
 //===----------------------------------------------------------------------===//
 // PackOp/UnPackOp Common
 //===----------------------------------------------------------------------===//
+// FIXME: Duplicates similar hook from TensorOps.cpp!
```
I've already removed one of these (`getNewMixedTileSizes` was no longer needed in `Tensor`). For more changes, I've extracted this pre-processing step:
* https://github.com/llvm/llvm-project/pull/126802

Once that lands, I will rebase this PR and `FIXME`s will be gone.

rengolin replied:
That PR looks good to me, thanks!

banach-space commented:
**Update (13/2/24)**

Re-based to resolve conflicts after https://github.com/llvm/llvm-project/pull/126802.

rengolin reviewed (approved):
Just rebased tpp-mlir on this PR and did not find any conflict relevant to the PR (only other changes that are also present in the commit where this PR is based on).

Code update was remarkably simple: 
* Code: `s/tensor::(Up)PackOp/linalg::(Un)PackOp/g`
* Tests: `s/tensor.(un)pack/linalg.(un)pack/g`

Thank you for the massive effort! Rebasing such a branch isn't easy, and since this is mainly a move, I'd be happy with it being merged sooner rather than later and we fix any teething issues in-tree.

banach-space committed:
fixup! [mlir][tensor][linalg] Move Pack/Unpack Ops to Linalg (4/4)

Update new test

banach-space commented:
Thanks, @rengolin! I really appreciate the reviews—both for this PR and all the dependent PRs!

@qedawkins, would you be able to allocate some time to review this in the coming week? Things are moving quickly in Linalg, and rebasing is starting to get a bit tricky.

qedawkins commented:
I am in the middle of rebasing the IREE side, will approve once it looks good. Had some trouble with my dev machine rebooting itself this morning so was a bit slow :(

qedawkins reviewed (approved):
Awesome, thanks @banach-space! Just a rename + updating some external interface registrations.

chsigg commented:
The tensor dialect now depends on the linalg dialect because of `linalg::TensorRelayoutOpInterface`. Doesn't this go against the dialect layering and that interface would better stay in the tensor dialect?
